src/controllers/memory-router/index.tsx

```diff
@@ -1,57 +1,73 @@
-import React from 'react';
+import React, { useRef } from 'react';
 
-import { createMemoryHistory, MemoryHistoryBuildOptions } from 'history';
+import { createMemoryHistory } from 'history';
 
 import { Router } from '../router';
-import { RouterProps } from '../router/types';
 
 import { MemoryRouterProps } from '../../common/types';
 
-const getRouterProps = (memoryRouterProps: MemoryRouterProps) => {
-  const {
-    isStatic = false,
-    isGlobal = true,
-    basePath,
-    routes,
-    resourceData,
-    resourceContext,
-  } = memoryRouterProps;
-  let routerProps: Partial<RouterProps> = {
-    basePath,
-    routes,
-    isStatic,
-    isGlobal,
+const lazy = <T extends any>(callback: () => T) => {
+  let firstCall = true;
+  let current: T | undefined = undefined;
+
+  return () => {
+    if (firstCall) {
+      current = callback();
+      firstCall = false;
+    }
+
+    return current;
   };
+};
 
-  if (resourceData) {
-    routerProps = { ...routerProps, resourceData };
-  }
+const useMemoryHistory = (location: string | undefined) => {
```

Comment:
Stored all the ugliness of trying to retain the same instance across renders in this hook.